lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -1116,6 +1118,228 @@ static bool canonicalizeOrOfConcatsWithCstOperands(OrOp op, size_t concatIdx1,
   return true;
 }
 
+namespace {
```

Comment:
https://llvm.org/doxygen/IntervalMap_8h.html
perhaps?

Reply:
Unfortunately, based on my understanding, intervals are merged only if the value associated is the same. Additionally, to plug in APInt as a key just as many wrappers are required. Since the location information should be merged for an interval, I don't see how adapting IntervalMap would be simpler.

Reply:
Two intervals which overlap or are adjacent should be merge with intervalmap.  This type of analysis on the llvm side is precisely what the datastructure was written for.

Reply:
https://github.com/nandor/llir-llvm/blob/master/llvm/unittests/ADT/IntervalMapTest.cpp#L214

I think this test asserts that intervals are not coalesced if the values they are mapped to are different. In this case, for each interval I also want to collect the source locations of all expressions contributing to it and I don't see how I can do that nicely with the interval map.

Reply:
Since I want the value to be a location, whenever a new range is inserted, the location of the expression testing for that range should be merged into the location attached to the whole interval in order to show that the test is now part of the comparison which will be emitted for the interval. Looking at the definition of `IntervalMap`, I think this step could be done by querying what overlaps with a new interval, computing the fusion of the locations and then updating the whole interval. Due to this overhead, I expect complexity to be worse than that of the sort and the implementation to be more convoluted. The `IntervalMap` might be great if queries and updates are interleaved, however in this case the another structure (sorted list) can be pre-processed before all the queries.